envoy/rds/config_traits.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+#include "envoy/rds/config.h"
+
+#include "source/common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Rds {
+
+/**
+ * Traits of the protocol specific route configuration and proto.
+ * The generic rds classes will call the methods of this interface
+ * to get information which are not visible for them directly.
+ */
+class ConfigTraits {
+public:
+  virtual ~ConfigTraits() = default;
+
+  /**
+   * Give the full name of the route configuration proto description.
+   * For example 'envoy.config.route.v3.RouteConfiguration'
+   */
+  virtual std::string resourceType() const PURE;
+
+  /**
+   * Create a dummy config object without actual route configuration.
+   * This object will be used before the first valid route configuration is fetched.
+   */
+  virtual ConfigConstSharedPtr createConfig() const PURE;
+
+  /**
+   * Create an empty route configuration proto object.
+   */
+  virtual ProtobufTypes::MessagePtr createProto() const PURE;
+
+  /**
+   * Runtime check if the provided proto message object is really a route configuration instance.
+   * Throw an std::bad_cast exception if not.
+   * Every other method below this assumes the proto message is already
+   * validated and doesn't do any further runtime check.
+   */
+  virtual const Protobuf::Message& validateResourceType(const Protobuf::Message& rc) const PURE;
```

Comment:
Why is validate returning a reference? Is this really an internal detail?

Reply:
The implementation for this is usually:
`return dynamic_cast<const envoy::config::route::v3::RouteConfiguration&>(rc);`
If no return value then the compiler fails because the unused result of dynamic cast.
Anyway you are right it is not needed. If you have any idea how to avoid this I change it.
I will also try to find some solution.

The validation is there in the original implementation as well in `RdsRouteConfigSubscription::onConfigUpdate`, just moved out behind this interface since the resource type is protocol specific.